docs/naps/8-multiple-canvases.md

```diff
@@ -0,0 +1,261 @@
+
+```{image} _static/multicanvas-napari-architecture-today.png
+---
+name: fig-1
+---
+Fig. 1: napari architecture today.
+```
+
+Figure 2 shows proposed changes (in orange) to the architecture to support multiple canvases. The new architecture is still following the MVC pattern. Again, this diagram only includes the Image layer type. Here is a summary of the planned changes:
+* Slice state will be moved off the layer as necessary, into new `LayerSlice` classes for each layer type
+    * This will be different for each Layer type - unifying the structure in the process may be a secondary benefit but is not the goal
+* Each VispyCanvas will hold a reference to a dedicated model class (`CanvasModel`)
+    * dims will move from `ViewerModel` -> `CanvasModel`
+    * camera will move from `ViewerModel` -> `CanvasModel`
+    * `layer_to_slice` will map each Layer (global list) to a `LayerSlice` (one per `CanvasModel`)
+* `ViewerModel` will own a *list* of `CanvasModel` objects
+* `QtViewer` will own a *list* of `VispyCanvas` objects
+* `VispyLayer` subclasses will hold references to their `Layer` (for rendering information) as well as a `LayerSlice` (for data)
+* The LayerSlicer will need to update the sync and async callbacks[^async-only] to pass a canvas parameter where the resulting sliced data will be stored, rather than storing it on the layer itself. Slice task cancellation logic will need to be revisited accordingly.
+    * async callback: `LayerSlicer._on_slice_done`
+    * sync callback: `Layer._slice_dims`
+* Callbacks (interaction, events) will need to be specifically connected to individual `CanvasModel` objects where relevant (dims, camera) rather than the `ViewerModel`.
+
+
+```{image} _static/multicanvas-napari-architecture-tomorrow.png
+---
+name: fig-2
+---
+Fig. 2: napari architecture tomorrow, with proposed changes from this NAP highlighted in orange.
+```
+
+> Note: in both diagrams, the `on_draw()` method on the `VispyCanvas` breaks MVC convention (view layer talks directly to the model layer). This is a separate/known issue and I believe is mostly only true for multiscale image layers right now. Changing this is considered out of scope for this project at this time.
+
+Slice state for each layer is currently stored on the Layer model. Again, see [NAP-4 for previous discussion](https://napari.org/stable/naps/4-async-slicing.html#existing-slice-state). This NAP proposes to move this state off the Layer instance, into a specific Layer Slice instance. This is what will allow multiple slices of a single layer to be visualized simultaneously. *Table 1* lists the attributes related to slice state that will be moved in this work from each Layer class into corresponding Layer Slice classes.
```

Comment:
Do you mean this non-goal:
> Normalizing slice data for different layer types, though this may benefit in the course of this work.

That's not really what I meant there, but also I think it's ambiguous enough warrant changing or removing. I was trying to say it is a non-goal to create a uniform/generic concept (or Protocol) of a "slice" - it's okay for the slices to be quite specific for each layer type.

Reply:
No, I meant this one.

> Separation of rendering information from the Layer(Data) model

But maybe you meant that only with respect to NAP-3.

Anyway, I also agree that a generic slice class/protocol should be a non-goal. Though we might consider during/after.

Reply:
I see - I will try to make this more clear. The goal here is indeed to move the slice state off the layer. The non-goal was meant to indicate that the remaining attributes of the layer will not be separated. So for example this work does not necessarily make it easier to share data between two layers with different colormaps.